File: utils/dnp3_src/cpp/libs/src/asiodnp3/DNP3Manager.cpp
/*
 * Licensed to Green Energy Corp (www.greenenergycorp.com) under one or
 * more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Green Energy Corp licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This project was forked on 01/01/2013 by Automatak, LLC and modifications
 * may have been made to this file. Automatak, LLC licenses these modifications
 * to you under the terms of the License.
 */
#include "asiodnp3/DNP3Manager.h"

#include "asiodnp3/DNP3ManagerImpl.h"

namespace asiodnp3
{

DNP3Manager::DNP3Manager(
    uint32_t concurrencyHint,
    std::shared_ptr<openpal::ILogHandler> handler,
    std::function<void()> onThreadStart,
    std::function<void()> onThreadExit) :
	impl(std::make_unique<DNP3ManagerImpl>(concurrencyHint, handler, onThreadStart, onThreadExit))
{

}

DNP3Manager::~DNP3Manager()
{}

void DNP3Manager::Shutdown()
{
	impl->Shutdown();
}

std::shared_ptr<IChannel> DNP3Manager::AddTCPClient(
    const std::string& id,
    uint32_t levels,
    const asiopal::ChannelRetry& retry,
    const std::string& host,
    const std::string& local,
    uint16_t port,
    std::shared_ptr<IChannelListener> listener)
{
	return this->impl->AddTCPClient(id, levels, retry, host, local, port, listener);
}

std::shared_ptr<IChannel> DNP3Manager::AddTCPServer(
    const std::string& id,
    uint32_t levels,
    const asiopal::ChannelRetry& retry,
    const std::string& endpoint,
    uint16_t port,
    std::shared_ptr<IChannelListener> listener)
{
	return this->impl->AddTCPServer(id, levels, retry, endpoint, port, listener);
}

std::shared_ptr<IChannel> DNP3Manager::AddSerial(
    const std::string& id,
    uint32_t levels,
    const asiopal::ChannelRetry& retry,
    asiopal::SerialSettings settings,
    std::shared_ptr<IChannelListener> listener)
{
	return this->impl->AddSerial(id, levels, retry, settings, listener);
}

std::shared_ptr<IChannel> DNP3Manager::AddTLSClient(
    const std::string& id,
    uint32_t levels,
    const asiopal::ChannelRetry& retry,
    const std::string& host,
    const std::string& local,
    uint16_t port,
    const asiopal::TLSConfig& config,
    std::shared_ptr<IChannelListener> listener,
    std::error_code& ec)
{
	return this->impl->AddTLSClient(id, levels, retry, host, local, port, config, listener, ec);
}

std::shared_ptr<IChannel> DNP3Manager::AddTLSServer(
    const std::string& id,
    uint32_t levels,
    const asiopal::ChannelRetry& retry,
    const std::string& endpoint,
    uint16_t port,
    const asiopal::TLSConfig& config,
    std::shared_ptr<IChannelListener> listener,
    std::error_code& ec)
{
	return this->impl->AddTLSServer(id, levels, retry, endpoint, port, config, listener, ec);
}

std::shared_ptr<asiopal::IListener> DNP3Manager::CreateListener(
    std::string loggerid,
    openpal::LogFilters loglevel,
    asiopal::IPEndpoint endpoint,
    std::shared_ptr<IListenCallbacks> callbacks,
    std::error_code& ec
)
{
	return impl->CreateListener(loggerid, loglevel, endpoint, callbacks, ec);
}

std::shared_ptr<asiopal::IListener> DNP3Manager::CreateListener(
    std::string loggerid,
    openpal::LogFilters loglevel,
    asiopal::IPEndpoint endpoint,
    const asiopal::TLSConfig& config,
    std::shared_ptr<IListenCallbacks> callbacks,
    std::error_code& ec
)
{
	return impl->CreateListener(loggerid, loglevel, endpoint, config, callbacks, ec);
}

}

File: utils/dnp3_src/java/bindings/src/main/java/com/automatak/dnp3/DNP3Manager.java
/**
 * Copyright 2013-2016 Automatak, LLC
 *
 * Licensed to Automatak, LLC (www.automatak.com) under one or more
 * contributor license agreements. See the NOTICE file distributed with this
 * work for additional information regarding copyright ownership. Automatak, LLC
 * licenses this file to you under the Apache License Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.automatak.dnp3;

/**
 * The main entry point for all dnp3 applications. Use this object to create communication channels to which
 * you can then add masters and outstations.
 *
 * To obtain a concrete implementation of this interface, use the DNP3ManagerFactory class in the impl package.
 *
 */
public interface DNP3Manager {

    /**
     * Add a TCP client channel. The channel does not try to connect until you add a stack.
     * @param id An id used for logging purposes
     * @param levels The starting level for logging output
     * @param retry Retry configuration for the channel
     * @param address The address of remote host as a INET address i.e. "127.0.0.1" or name "www.google.com"
     * @param adapter The local adapter to use. "0.0.0.0* means "any adapter'.
     * @param port The port to make the connection on. Note that only the range 0 to 65535 is valid
     * @param listener Optional listener (can be null) for monitoring the state of the channel
     * @return A channel interface
     * @throws DNP3Exception if an error occurs creating the channel
     */
    Channel addTCPClient(String id, int levels, ChannelRetry retry, String address, String adapter, int port, ChannelListener listener) throws DNP3Exception;

    /**
     * Add a TCP client channel. The channel does not try to connect until you add a stack.
     * @param id An id used for logging purposes
     * @param levels The starting level for logging output
     * @param retry Retry configuration for the channel
     * @param endpoint TThe address that identifies the network adapter to bind i.e. "127.0.0.1" or "0.0.0.0"
     * @param port The port to make the connection on. Note that only the range 0 to 65535 is valid
     * @param listener Optional listener (can be null) for monitoring the state of the channel
     * @return A channel interface
     * @throws DNP3Exception if an error occurs creating the channel
     */
    Channel addTCPServer(String id, int levels, ChannelRetry retry, String endpoint, int port, ChannelListener listener) throws DNP3Exception;

    /**
     * Add a TCP client channel. The channel does not try to connect until you add a stack.
     * @param id An id used for logging purposes
     * @param levels The starting level for logging output
     * @param retry Retry configuration for the channel
     * @param address The address of remote host as a INET address i.e. "127.0.0.1" or name "www.google.com"
     * @param adapter The local adapter to use. "0.0.0.0* means "any adapter'.
     * @param port The port to make the connection on. Note that only the range 0 to 65535 is valid
     * @param config TLS configuration
     * @param listener Optional listener (can be null) for monitoring the state of the channel
     * @return A channel interface
     * @throws DNP3Exception if an error occurs creating the channel
     */
    Channel addTLSClient(String id, int levels, ChannelRetry retry, String address, String adapter, int port, TLSConfig config, ChannelListener listener) throws DNP3Exception;

    /**
     * Add a TCP client channel. The channel does not try to connect until you add a stack.
     * @param id An id used for logging purposes
     * @param levels The starting level for logging output
     * @param retry Retry configuration for the channel
     * @param endpoint TThe address that identifies the network adapter to bind i.e. "127.0.0.1" or "0.0.0.0"
     * @param port The port to make the connection on. Note that only the range 0 to 65535 is valid
     * @param config TLS configuration
     * @param listener Optional listener (can be null) for monitoring the state of the channel
     * @return A channel interface
     * @throws DNP3Exception if an error occurs creating the channel
     */
    Channel addTLSServer(String id, int levels, ChannelRetry retry, String endpoint, int port, TLSConfig config, ChannelListener listener) throws DNP3Exception;

    /**
     * Add a serial channel. The port does not try to open until you add a stack.
     *
     * @param id An id used for logging purposes
     * @param levels The starting level for logging output
     * @param retry Retry configuration for the channel
     * @param settings Configuration for the serial port
     * @param listener Optional listener (can be null) for monitoring the state of the channel
     * @throws DNP3Exception if an error occurs creating the channel
     * @return a channel interface
     */
    Channel addSerial(String id, int levels, ChannelRetry retry, SerialSettings settings, ChannelListener listener) throws DNP3Exception;


    /**
     * Permanently shutdown all channels and any associated stacks
     */
    void shutdown();
}
